### PR TITLE
Mock the plug-in modules and refactor the VM context API.

### PR DESCRIPTION
**Pull request by q82419**

### CODE REVIEW

**Comment by q82419** [comment]
@apepkuss Please notice this PR that the API changes. Thanks.

**Comment by apepkuss** [comment]
@q82419 Thank you for letting me know about the changes. I checked the changes on `vm.rs` and found that only the API names are updated but no semantics get changed, right? Thanks!

**Comment by q82419** [comment]
> @q82419 Thank you for letting me know about the changes. I checked the changes on `vm.rs` and found that only the API names are updated but no semantics get changed, right? Thanks!

Yes, only the name changes.
But for the `ModuleInstance`, several API deleted and replaced by the `Plugin` APIs.

**Comment by apepkuss** [comment]
> > @q82419 Thank you for letting me know about the changes. I checked the changes on `vm.rs` and found that only the API names are updated but no semantics get changed, right? Thanks!
> 
> Yes, only the name changes. But for the `ModuleInstance`, several API deleted and replaced by the `Plugin` APIs. I'll update the release note and C API docs these days. You can check that then.

Got it. Thanks a lot!

**Comment by q82419** [comment]
Hi @kenvifire ,

Would you want to help to update the Java SDK for the API changes after this PR being merged?
Thanks!

**Comment by kenvifire** [comment]
> Hi @kenvifire ,
> 
> Would you want to help to update the Java SDK for the API changes after this PR being merged? Thanks!

Sure.

File: include/host/mock/wasi_crypto_func.h [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+namespace WasiCryptoMock {
+
+using namespace std::literals;
+
+namespace Common {
+class ArrayOutputLen : public Runtime::HostFunction<ArrayOutputLen> {
+public:
+  Expect<uint32_t> body(const Runtime::CallingFrame &, int32_t, uint32_t) {
+    printPluginMock("WASI-Crypto"sv);
+    return 1U;
```
Comment (hydai) [inline include/host/mock/wasi_crypto_func.h:22]: Please use a constant error code to replace this magic number.

File: bindings/java/wasmedge-java/wasmedge-jni/lib/WasmEdgeVM.c [hunk 2]
```diff
@@ -553,7 +553,7 @@ Java_org_wasmedge_WasmEdgeVm_nativeGetImportModuleContext(JNIEnv *env,
                                                           jobject thisObject,
                                                           jint reg) {
   WasmEdge_VMContext *vmContext = getVmContext(env, thisObject);
-  WasmEdge_ModuleInstanceContext *imp = WasmEdge_VMGetImportModuleContext(
+  WasmEdge_ModuleInstanceContext *imp = WasmEdge_VMGetPreRegisteredModule(
```
Comment (hydai) [inline bindings/java/wasmedge-java/wasmedge-jni/lib/WasmEdgeVM.c:556]: This rename is not necessary; please undo this.

File: include/host/mock/wasi_nn_func.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+namespace WasiNNMock {
+
+using namespace std::literals;
+
+class Load : public Runtime::HostFunction<Load> {
+public:
+  Expect<uint32_t> body(const Runtime::CallingFrame &, uint32_t, uint32_t,
+                        uint32_t, uint32_t, uint32_t) {
+    printPluginMock("WASI-NN"sv);
+    return 1U;
```
Comment (hydai) [inline include/host/mock/wasi_nn_func.h:22]: Use constant error code.

File: lib/vm/vm.cpp [hunk 4]
```diff
@@ -27,64 +48,68 @@ VM::VM(const Configure &Conf, Runtime::StoreManager &S)
...
+
+  // Load the other non-official plugins.
   for (const auto &Plugin : Plugin::Plugin::plugins()) {
     if (Conf.isForbiddenPlugins(Plugin.name())) {
       continue;
     }
-    // skip WasmEdge_Process, wasi_nn, and wasi_crypto.
-    if (Plugin.name() == "wasmedge_process"sv || Plugin.name() == "wasi_nn"sv ||
-        Plugin.name() == "wasi_crypto"sv) {
+    // Skip WasmEdge_Process, WasmEdge_HttpsReq, wasi_nn, and wasi_crypto.
```
Comment (hydai) [inline lib/vm/vm.cpp:85]: extract an external function for this section.

**Review by hydai — CHANGES_REQUESTED** [review]

**Review by hydai — APPROVED** [review]

**Comment by apepkuss** [comment]
@q82419 Please ignore the failures caused by the workflows related to Rust SDK. I'll fix them in another PR. Thanks!

**Pull request merged**